### ISSUE TITLE
events 

### ISSUE THREAD

**Issue by magicwang-cn** [kubernetes/kubernetes#19637]
currently, events are stored into etcd, i think this is not a good way. because events always multi-times than pods, if the cluster is big(`15000 pods, there maybe  > 100K events`), etcd can not handle it effectively. Than the e2e test(`density test`) maybe not works fine.
why not send the events to a manager center(`like kafka`), or store them in the database?

**Comment by Pensu** [comment]
Not sure about etcd's capability. Has anyone actually tried handling >100K events using etcd?

**Comment by gmarek** [comment]
We were talking about it some time ago, but we didn't feel that it has high priority. @wojtek-t @kubernetes/sig-scalability

**Comment by magicwang-cn** [comment]
@Pensu @gmarek 
i have a test with 500 nodes and 15000 pods(`e2e density test`), there are more than 60K events,  and the e2e density's Latency test will not works fine, it is slowly or badly to [list schedule events](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/density.go#L355).

>Not sure about etcd's capability. Has anyone actually tried handling >100K events using etcd?

**Comment by gmarek** [comment]
Yeah, but that's test only thing. Events are not really first-class citizen, as they are designed for human consumption and no component should depend on them. For this reason we, as a scalability team care for them only as much as to prevent them from impacting performance of other systems.

For this reason first thing I would do wouldn't be changing backend, but spinning up a new API server to handle Events only.

**Comment by timothysc** [comment]
> why not send the events to a manager center(like kafka), or store them in the database?

@magicwang-cn +10k, but you would need to support the storage interface and configure the shard the api-server.  As of today, you can shard events to a separate etcd.

**Comment by timothysc** [comment]
@mikedanese please add label area-performance for tracking.

**Comment by magicwang-cn** [comment]
@timothysc yes. i have already shard events to a separate etcd, but in large cluster, it is works bad. so maybe store events to database or kafka is a better way.
anyone else do it?

**Comment by magicwang-cn** [comment]
@alfred-huangjian @HardySimpson @kevin-wangzefeng

**Comment by jayunit100** [comment]
I'm pretty sure Kafka is indeed a better way, the write througput should work much better for an event stream, because I think the node receiving the writes isn't frozen. So the durability guarantee is cheaper than the strong type of key level consistency etcd provides, which forces all writes to go through the same node.

**Comment by timothysc** [comment]
@magicwang-cn If you are interested in writing a storage wrapper for kafka I'd be happy to give an overview at the scalability-SIG meeting, and give an update re: #18542

**Comment by gmarek** [comment]
cc @mwielgus @piosz

**Comment by alfred-huangjian** [comment]
@timothysc
As far as I know, Heapster currently watch the api server and collect the events also. But for events, the heapster only support influxDB sink now. It will be good to have an alternative choice to support sending the events directly to kafka instead of storing into etcd. Kafka can be easily to used to connected some third-party ops system. For example, heapster can collect the events from the kafka.

@huangyuqi has worked on the kafka sink in heapster. I think here @magicwang-cn and @huangyuqi can handle this task here.

**Comment by alfred-huangjian** [comment]
cc @kevin-wangzefeng @HardySimpson

**Comment by magicwang-cn** [comment]
@timothysc yes. I am happy to do this work. any suggestion?

**Comment by timothysc** [comment]
@magicwang-cn 

Step 1: introduce a storage that stubs and implements as much of the interface as possible.  
https://github.com/kubernetes/kubernetes/blob/master/pkg/storage/interfaces.go

**Comment by lavalamp** [comment]
Yeah, listing 100k events is going to be painful no matter the storage backend, because we don't have paging.

**Comment by lavalamp** [comment]
I'd be very happy to take a PR adding an alternative storage destination.

**Comment by magicwang-cn** [comment]
@lavalamp you mean you will do this work?

**Comment by lavalamp** [comment]
No, I mean I'd be happy to review such a pr-- I don't have time to make one personally unfortunately.

**Comment by magicwang-cn** [comment]
@lavalamp ok. :)

**Comment by davidopp** [comment]
@kubernetes/huawei

**Comment by magicwang-cn** [comment]
@timothysc @lavalamp 
recently, i have do a research of kafka, it is not a normal storage system, it has no update/delete interface, really a publish-subscribe system.

**kafka:**

- only topic to separate different messages, not a key-value storage
  - just one topic "**events**" to subscribe all namespaces' events, but this will reduce the read action's efficiency.
  - one namespace one topic, like resource-namespace(`events-default`), but once the namespace is very much, maybe too many topics is not a good use of kafka
- it has no update/delete/watch interface
  - update is just the same as create, fortunately, events do not need update
  - delete is the kafka server's action, you can set the expired-time configuration of the events, but it not support with api.
  - is aslo not support watch interface, we can complete it in the apiserver side(`like the list-watch mechanism`), but it will make apiserver more complex, is it a good way?

so any advice will be ok, thanks.

**Comment by jiangyaoguo** [comment]
@magicwang-cn Kubernetes will update events for event compression. So we need consider update interface or change event mechanism. Now event is more like log rather than reliable message. You can see more discussion in #4432.  Perhaps we can reuse a separate apiserver which handle event requests only and abstract interfaces which can be implemented by MongoDB and so on.

**Comment by magicwang-cn** [comment]
@jiangyaoguo yes. the [storage interface](https://github.com/kubernetes/kubernetes/blob/master/pkg/storage/interfaces.go) is not suitable for events. we just want create/read actions. so the events interface should be different from storage interface.

**Comment by mikedanese** [comment]
We might be able to get rid of compression if we move to a proper storage for events.

**Comment by chancez** [comment]
My current plan for this is to write a simple pod which watches for namespaces being created/removed, and then for each namespace watches incoming events and logs them to stdout. By doing this, the pod logs will be picked up by fluentd and stored into elasticsearch. It should work as a simple work-around until a better solution is created.

This will enable searching/indexing events however you want, and doesn't require much effort. Will try to update here once I get something working.

**Comment by mwielgus** [comment]
@chancez We have a Heapster component called Eventer (https://github.com/kubernetes/heapster/tree/master/events) that watches Events on the master and writes them to one of the permanent storages. There is a pending PR with ElasticSearch support.

Please consider using this component.

**Comment by chancez** [comment]
@mwielgus Thanks for pointing this out. The log sink would be sufficient since heapster's logs should get picked up by fluentd anyways, so I don't even know that having an ES sink is all that important right now.

It looks like the `deploy` folder in heapster is pretty out of date, is the heapster stuff from cluster/addons in the kubernetes repo sufficient to have this functionality?

**Comment by timothysc** [comment]
Perhaps the easiest way to offload out of the main api-server would be to proxy /events to an introspective service running an api-server+etcd as a pod, provided the service exists otherwise default to original behavior.  

Problem with complete offload (a.k.a kafaka) today, is that we would need to proliferate connection+sec details to all clients + update all backing client init functions... 

/cc @rrati

**Comment by falenn** [comment]
First of all Kubernetes is awesome - great work!

The last post here is somewhat old - any movement on this?

The last thing we would want in a production cluster is back-pressure from eventing to crush the cluster.  I would think events need to drain from the system via asynchronous, non-blocking mechanisms.  Message brokers such as Rabbit, Kafka, ZMQ, S3 and the like are built for this purpose, providing an integration point down-stream for options on how to deal with the events, fanning out their destinations, choosing different storage or analytical targets, etc.  I'd be in favor of a configurable eventing posture in order to avoid pressuring Etcd.  
@timothysc You're right that this may complicate some things - perhaps an alternative config so not part of basic setup?

**Comment by timothysc** [comment]
I've started https://github.com/heptio/eventrouter to push events to other systems.

**Comment by piosz** [comment]
@timothysc how does it differ from [eventer](https://github.com/kubernetes/heapster/tree/master/events)?

**Comment by timothysc** [comment]
@piosz it's very similar, if I had known prior to starting that it was there I would have submitted patches, but TBH I'm kind of glad I made it separate.  I will likely evolve it independently per our own requirements, and it's small enough that it doesn't matter much.

**Comment by JohnRusk** [comment]
I see this has been around for a while. (Approaching 8 years).  As far as I can tell, it's still an issue that exists.  What's the current thinking on what the direction should be?
